### DISCUSSION TITLE
Project management

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#3665] [General]
I'm struggling a bit with the project management side of `rust-bitcoin`. We have too many things going on to keep in my head these days. I've been using issues and listing PRs/issues + random TODOs (eg #3133).

We have in the past tried using GitHub projects, milestones, and org level projects - none with much success.

The current task at hand is get the 1.0 out for `units`, `io`, `hashes`, `hex`, `ordered`, and `primitives` - note 2 of these are not in this repo.

Is there a better way to do the project management.

**Comment by dpc** [comment]
BTW. Did I ever shill [The Goal](https://en.wikipedia.org/wiki/The_Goal_%28novel%29) here? I shill it to everyone who asks anything related, as it so well describes my conclusions to both: software project management and distributed system optimization, while being a book about something else anyway: manufacturing management.

In my experience adding meta-work rarely ever makes work go faster, quite the opposite. Project management is like locking primitives - often the necessary overhead to distribute and coordinate concurrent, interdependent work. There is a lot of aspects to it, as it's really a distributed system optimization task, but generally keeping it lightweight and avoiding contention and inter-dependencies is common sense rule of thumb.


I don't think there's that many developers in rust-bitcoin for the coordation to be an issue. "too many things going on to keep in my head" sounds like working set being too large for a single (few) workers, causing poor cache utilization. The solution is narrowing the working set, focusing on one smaller step that fits in the cache at the time and driving it to completion, avoiding context switches.

Context switches itself are usually caused by high latencies. When a task is blocked and can't be progressed, it seems like it makes sense to switch to something else to increase utilization. People have a natural tendency to want to optimize throughput, which is often counter-productive, because context switches are more expensive than it seems. The Goal in great length describes reason for and benefits of "minimizing work in progress". In rust-bitcoin my bet would be that latencies are caused by slow review and merging pipeline. You make a PR, you need reviews and approvals, you switch to something else - context switch, and another and another, eventually you lost track of what's going on with some of the work. I'm not sure how much I can help because it's a cultural and worldview issue - wanting to be deliberate and thorough. But it is something to think about.

If you e.g. can have someone dedicated to providing low-latency PR reviews and other feedback, that might increase the iteration time and minimize amount of work in progress. Or if debates and lenghty back and forth during reviews are common, why not schedule a dedicated time for multiple people to do a review and discuss opinions over a video call? Countless times I notice in all projects the pattern of: someone makes a PR, waits a day for a review, takes a day to address changes, waits a day to get another round of feedback, someone else chims in, another day to address feedback, and so on. If they all jumped on a call together they might have been done in 30 minutes.

Anyway, the general way to go about it is to make a mental note every time you need to context switch and think about what could have been done to avoid it.

Just prioritzing completing existing work over starting new one, is a generally good approach.

Another approach to eliminate latency is pipelining: could you somehow finish your work, without needing to get back to it? E.g. in Fedimint we have a tag, where after a single review developes can land a PR and mark it as "LGTM, but someone should go and check on it even after it lands". This to minimize the time things are stuck waiting, a form of optimistic locking. `git revert` is there for a reason.

Another approach to lowering latency is batching: line up N PRs, then have a video call with someone to review and discuss all of them without extra reviewing round-trips.

Related: [Iteration time trumps everything](https://dpc.pw/posts/iteration-time-trumps-everything/)

**Comment by tcharding** [comment]
Epic, thanks man. I ordered your book.

**Comment by storopoli** [comment]
I think that a VERY simple approach with GitHub projects and the [board layout](https://docs.github.com/en/issues/planning-and-tracking-with-projects/customizing-views-in-your-project/customizing-the-board-layout) might be an interesting solution.

![image](https://github.com/user-attachments/assets/5fbacf2c-b83a-42fd-8a1a-cf312929e60d)

The board layout (pictured above) can be highly-customizable with all the "agile"/"kan ban"/"scrum master" crap. But it can be also left very barebones.
And that's what I think should be tried first.
Just a simple view with maybe 3 buckets: TODO, IN PROGRESS, IN REVIEW.

You can also add milestones tied to releases and also create different boards for each one of the crates `units`, `io`, `hashes`, `hex`, `ordered`, and `primitives` etc. Even if they are in different repos, as you pointed out, you can create an organization-level (`rust-bitcoin` org) GitHub project to track these.

This is an interesting solution since it has some automation automatically baked in, such as if you add an issue or PR to the project it will be shown in it and once closed or merged will be removed as well (or marked completed). It is also possible to cross-ref and link with anything in GitHub, such as issues, discussions, PRs, commits, CI etc.
It is a way better approach with not much overhead for your

> using issues and listing PRs/issues + random TODOs

Just don't try to be fancy with GH Projects (yes, you can if you want). Keep it simple as a dynamic TODO view of stuff that needs to be kept in your radar.

**Comment by yancyribbens** [comment]
> The current task at hand is get the 1.0 out for units, io, hashes, hex, ordered, and primitives - note 2 of these are not in this repo.

Out of these 6 crates you listed, which is the easiest/closest to being in a 1.0 state?  What are the remaining TODOs to get that crate complete?

**Reply by tcharding** [reply]
https://github.com/rust-bitcoin/rust-bitcoin/issues?q=is%3Aissue+is%3Aopen+release+tracking

**Reply by yancyribbens** [reply]
This looks well organized and like you have things identified that need priority.  I guess I'm a little confused what the issue is your struggling with.

**Reply by yancyribbens** [reply]
s/your/you're.  Some day I'll learn English.

**Comment by tcharding** [comment]
@dpc answered all my concerns.
